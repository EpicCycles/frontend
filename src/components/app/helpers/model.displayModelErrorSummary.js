import {customerAddressFields} from "../model/helpers/fields";
import {displayModelErrorSummary} from "../model/helpers/model";

test('if no error or error_detail undefined returned', () => {
    const model = {
        id: 123,
        address1: "line one",
        address2: "line Ywo",
        address3: "line Three",
        address4: "line Four",
        postcode: "xxxyyy",
        customer: 6
    };
    const result = displayModelErrorSummary(model, customerAddressFields);
    expect(result).toBe(undefined);
});
test('if just error no error_detail error returned', () => {
    const model = {
        id: 123,
        address1: "line one",
        address2: "line Ywo",
        address3: "line Three",
        address4: "line Four",
        postcode: "xxxyyy",
        customer: 6,
        error: "Model level error"
    };
    const result = displayModelErrorSummary(model, customerAddressFields);
    expect(result).toBe("Model level error");
});
test('if no error with error_detail error_details formatted returned', () => {
    const model = {
        id: 123,
        address1: "line one",
        address2: "line Ywo",
        address3: "line Three",
        address4: "line Four",
        postcode: "xxxyyy",
        customer: 6,
        error_detail: {
            "address1": ["A valid address is required.", "SHow me the address is required."],
            "postcode": ["This field may not be blank."]
        },
    };
    const expectedText = "Address: A valid address is required. SHow me the address is required.<br>Postcode: This field may not be blank.";
    const result = displayModelErrorSummary(model, customerAddressFields);
    expect(result).toBe(expectedText);
});
test('if error and error_detail formatted returned', () => {
    const model = {
        id: 123,
        address1: "line one",
        address2: "line Ywo",
        address3: "line Three",
        address4: "line Four",
        postcode: "xxxyyy",
        customer: 6,
        error: "Model level error",
        error_detail: {
            "address1": ["A valid address is required.", "SHow me the address is required."],
            "postcode": ["This field may not be blank."]
        },
    };
    const expectedText = "Model level error<br>Address: A valid address is required. SHow me the address is required.<br>Postcode: This field may not be blank.";
    const result = displayModelErrorSummary(model, customerAddressFields);
    expect(result).toBe(expectedText);
});