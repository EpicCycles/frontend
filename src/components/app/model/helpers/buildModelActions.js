/**
 * buildModelActions - builds actions for the current model (use to build additional actions)
 * @param requiredActions - array of objects {iconName, iconTitle, requiresIdOnly, actionFunction}
 * @param model - model being displayed oredited
 * @returns array of model actions specific to current model
 */
export const buildModelActions = (requiredActions, model) => {
  return requiredActions.map(action => {
    const finalAction = {
      iconName: action.iconName,
      iconTitle: action.iconTitle,
    };
    const isSavedModel = model && model.id;
    if (action.requiresIdOnly) {
      finalAction.iconAction = () => isSavedModel && action.actionFunction(model.id);
    } else {
      finalAction.iconAction = () => isSavedModel && action.actionFunction(model);
    }
    finalAction.iconDisabled = !isSavedModel;
    return finalAction;
  });
};
