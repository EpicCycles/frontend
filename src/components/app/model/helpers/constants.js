export const POSTCODE_RULES = [
  { countryCode: 'AT', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'AU', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'BE', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'BG', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'CA', regex: /^[A-Za-z][0-9][A-Za-z] [0-9][A-Za-z][0-9]/, display: 'ANA NAN' },
  { countryCode: 'CH', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'CY', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'CZ', regex: /^\d{3}\s\d{2}$/, display: 'NNN NN' },
  { countryCode: 'DE', regex: /^\d{5}$/, display: 'NNNNN' },
  { countryCode: 'DK', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'EE', regex: /^\d{5}$/, display: 'NNNNN' },
  { countryCode: 'ES', regex: /^\d{5}$/, display: 'NNNNN' },
  { countryCode: 'FI', regex: /^\d{5}$/, display: 'NNNNN' },
  { countryCode: 'FR', regex: /^\d{5}$/, display: 'NNNNN' },
  {
    countryCode: 'GB',
    regex: /^[A-Z](([0-9][0-9]?)|([A-Z][0-9][0-9]?)|([A-Z]?[0-9][A-Z])) ?[0-9][ABD-HJLNP-UW-Z]{2}$/,
    display: 'CCNN NCC',
  },
  { countryCode: 'GR', regex: /^\d{3}\s\d{2}$/, display: 'NNNNN or NNN NN' },
  { countryCode: 'HU', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'IT', regex: /^\d{5}$/, display: 'NNNNN' },
  { countryCode: 'KR', regex: /^\d{5}$/, display: 'NNNNN' },
  { countryCode: 'LT', regex: /^\d{5}$/, display: 'NNNNN' },
  { countryCode: 'LU', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'MT', regex: /^\[A-Z]{3}\s\d{4}$/, display: 'AAA NNNN' },
  { countryCode: 'MY', regex: /^\d{5}$/, display: 'NNNNN' },
  { countryCode: 'NO', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'NZ', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'PH', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'PL', regex: /^\d{2}-\d{3}$/, display: 'NN-NNN' },
  { countryCode: 'PT', regex: /^\d{4}-\d{3}$/, display: 'NNNN-NNN' },
  { countryCode: 'RO', regex: /^\d{6}$/, display: 'NNNNNN' },
  { countryCode: 'SE', regex: /^\d{3,4}$/, display: 'NNNNN or NNNN' },
  { countryCode: 'SG', regex: /^\d{6}$/, display: 'NNNNNN' },
  { countryCode: 'SI', regex: /^\d{3}\s\d{2}$/, display: 'NNN NN' },
  { countryCode: 'SK', regex: /^\d{3}\s\d{2}$/, display: 'NNN NN' },
  { countryCode: 'US', regex: /^(\d{5}([-]\d{4})?)$/, display: 'NNNNN-NNNN or NNNNN' },
  { countryCode: 'ZA', regex: /^\d{4}$/, display: 'NNNN' },
];
