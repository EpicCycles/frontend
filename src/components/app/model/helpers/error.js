export const BRAND_NAME_MISSING = "A name is required for the brand";
export const FRAME_NAME_MISSING = "A name is required for the frame";
export const ATTRIBUTE_NAME_MISSING = "A name is required for the attribute";
export const BUNDLE_NAME_MISSING = "A name is required for the bundle";
export const MODEL_NAME_MISSING = "A name is required for the model";
export const PART_NAME_MISSING = "A name is required for the part";
export const PART_TYPE_NAME_MISSING = "A name is required for the part type";
export const SECTION_MISSING = "A section must be selected for the part type";
export const PART_TYPE_MISSING = "A part type must be selected";
export const BRAND_MISSING = "A brand must be selected";
export const PRODUCTS_MISSING = "At least 2 products must be selected";
export const SUPPLIER_MISSING = "A supplier must be selected";
export const PART_MISSING = "A part must be selected";
export const ADDRESS_MISSING = "At least one line of address must be entered";
export const COUNTRY_MISSING = "A country must be selected";
export const INVALID_POSTCODE = "A postcode is expected which for this country has the format ";
export const SELECT_ONE_MISSING = "Please select an option.";
export const VALUE_MISSING = "A value is required.";
export const INVALID_EMAIL = "Invalid email address.";
export const INVALID_URL = "Invalid URL";
