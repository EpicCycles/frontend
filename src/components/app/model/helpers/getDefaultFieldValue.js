export const getDefaultFieldValue = modelField => {
  if (modelField.selectList && Array.isArray(modelField.selectList)) {
    const defaultValue = modelField.selectList.filter(listitem => listitem.isDefault);
    if (defaultValue.length > 0) return defaultValue[0].value;
  }
  return modelField.default;
};
