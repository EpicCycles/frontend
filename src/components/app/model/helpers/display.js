import {
  BIKE,
  BRAND,
  CHECKBOX,
  CURRENCY,
  CUSTOMER,
  DATE_TIME,
  NUMBER,
  PART_TYPE,
  RADIO,
  SELECT_ONE,
  SUPPLIER,
  USER,
} from './fields';
import { getCustomerName } from '../../../customer/helpers/customer';
import { getBrandName } from '../../../brand/helpers/brand';
import { getNameForValue } from './model';
import { getSupplierName } from '../../../supplier/helpers/supplier';
import { getPartTypeName } from '../../../partType/helpers/partType';
import { getBikeName } from '../../../bike/helpers/bike';
import { getUserName } from '../../../user/helpers/user';
import { CHARGE } from '../../../quoteCharge/helpers/quoteChargeFields';
import { chargeName } from '../../../charge/helpers/chargeName';
import { findObjectWithKey } from '../../../../helpers/utils';
import { fittingText } from '../../../fitting/helpers/fitting';
import { FITTING } from '../../../quote/helpers/quoteFields';

export const fixedHeaderClassname = lockColumn => {
  if (lockColumn) return 'grid-header--fixed-left';
  return '';
};
export const fixedDetailsClassname = lockColumn => {
  if (lockColumn) return 'grid-item--fixed-left';
  return '';
};
export const gridHeaderClass = (baseClassName = '', fieldIndex, firstColumnLocked) => {
  const shouldLock = firstColumnLocked && fieldIndex === 0;
  return `${baseClassName} grid-item--header ${fixedHeaderClassname(shouldLock)}`;
};
export const gridItemClass = (baseClassName = '', fieldIndex, firstColumnLocked) => {
  const shouldLock = firstColumnLocked && fieldIndex === 0;
  return `grid-item ${fixedDetailsClassname(shouldLock)} ${baseClassName}`;
};
export const formattedDate = date => {
  if (date) return date.toLocaleDateString('en-GB');
  return '';
};
export const formattedDateTime = date => {
  if (date) return date.toLocaleString('en-GB');
  return '';
};
export const formattedPrice = fieldValue => {
  if (fieldValue)
    return Number(fieldValue).toLocaleString('en-GB', {
      style: 'currency',
      currency: 'GBP',
    });
  return '';
};
export const fieldAlignment = field => {
  switch (field.type) {
    case NUMBER:
    case CURRENCY:
      return 'align_right';
    case CHECKBOX:
    case RADIO:
      return 'align_center';
    default:
      return '';
  }
};
export const buildViewString = (model, field, sourceDataArrays = {}) => {
  let viewData;
  const {
    sections,
    brands,
    suppliers,
    customers,
    bikes,
    frames,
    users,
    charges,
    fittings,
  } = sourceDataArrays;
  const fieldValue = model ? model[field.fieldName] : undefined;
  switch (field.type) {
    case CUSTOMER:
      viewData = fieldValue ? getCustomerName(fieldValue, customers) : '';
      break;
    case CURRENCY:
      viewData = formattedPrice(fieldValue);
      break;
    case CHECKBOX:
      viewData = fieldValue ? 'Yes' : 'No';
      break;
    case DATE_TIME:
      viewData = fieldValue ? formattedDateTime(new Date(fieldValue)) : '';
      break;
    case PART_TYPE:
      viewData = fieldValue ? getPartTypeName(fieldValue, sections) : '';
      break;
    case BRAND:
      viewData = fieldValue ? getBrandName(fieldValue, brands) : '';
      break;
    case SUPPLIER:
      viewData = fieldValue ? getSupplierName(fieldValue, suppliers) : '';
      break;
    case BIKE:
      viewData = fieldValue ? getBikeName(fieldValue, bikes, frames, brands) : '';
      break;
    case SELECT_ONE:
      viewData = fieldValue ? getNameForValue(fieldValue, field.selectList) : '';
      break;
    case USER:
      viewData = fieldValue ? getUserName(fieldValue, users) : '';
      break;
    case CHARGE:
      viewData = fieldValue ? chargeName(fieldValue, charges) : '';
      break;
    case FITTING:
      const fitting = fieldValue ? findObjectWithKey(fittings, fieldValue) : undefined;
      viewData = fitting ? fittingText(fitting) : '';
      break;
    default:
      viewData = fieldValue ? fieldValue : '';
  }
  return viewData;
};
