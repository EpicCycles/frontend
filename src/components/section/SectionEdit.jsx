import React, { Fragment } from 'react';
import FormTextInput from '../../common/FormTextInput';
import { generateRandomCode, updateObject } from '../../helpers/utils';
import PartTypes from '../partType/PartTypes';
import { Icon } from 'semantic-ui-react';
import { NEW_ELEMENT_ID } from '../../helpers/constants';
import AddLink from "../app/model/AddLink";

class SectionEdit extends React.Component {
  handleInputChange = (fieldName, input) => {
    const updatedSection = updateObject(this.props.section);
    if (fieldName.startsWith('name')) updatedSection.name = input;
    if (fieldName.startsWith('partTypes')) updatedSection.partTypes = input;
    if (fieldName.startsWith('detail')) updatedSection._detail = input;
    if (!updatedSection.name) {
      updatedSection.error = true;
      updatedSection.error_detail = 'A name is required for the section';
    } else {
      updatedSection.error = false;
      updatedSection.error_detail = '';
    }

    let componentKey = this.props.componentKey;
    if (componentKey === NEW_ELEMENT_ID) {
      updatedSection.dummyKey = NEW_ELEMENT_ID;
    }
    updatedSection.changed = true;
    this.props.handleSectionChange(componentKey, updatedSection);
  };

  handleInputClear = fieldName => {
    if (
      window.confirm(
        'Please confirm that you want to delete this section and all associated part types and their attributes.',
      )
    ) {
      const updatedSection = updateObject(this.props.section);
      updatedSection.delete = true;
      this.props.handleSectionChange(this.props.componentKey, updatedSection);
    }
  };

  toggleDetail = () => {
    this.handleInputChange('detail', !this.props.section._detail);
  };
  addAnother = () => {
    const updatedSection = updateObject(this.props.section);
    updatedSection.dummyKey = generateRandomCode();
    this.props.handleSectionChange(NEW_ELEMENT_ID, updatedSection);
  };

  render() {
    const { section, componentKey } = this.props;
    const partTypes = section.partTypes || [];
    return (
      <Fragment>
        <div className="grid-item grid-item--fixed-left">
          {componentKey !== NEW_ELEMENT_ID ? (
            <Icon
              name={`toggle ${section._detail ? 'down' : 'right'}`}
              onClick={this.toggleDetail}
            />
          ) : (
            <AddLink addFunction={this.addAnother} addObjectName={'Section'} />
          )}
        </div>
        <div className="grid-item">
          <FormTextInput
            placeholder="add new"
            fieldName={`name_${componentKey}`}
            value={section.name}
            onChange={this.handleInputChange}
            onClick={this.handleInputClear}
          />
          {section._detail && componentKey !== NEW_ELEMENT_ID && (
            <PartTypes
              sectionKey={componentKey}
              partTypes={partTypes}
              handleSectionChange={this.handleInputChange}
            />
          )}
        </div>
      </Fragment>
    );
  }
}

export default SectionEdit;
