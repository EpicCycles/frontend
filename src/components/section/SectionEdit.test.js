import { NEW_ELEMENT_ID } from '../../helpers/constants';
import SectionEdit from './SectionEdit';

test('SectionEdit shows a NEW section correctly', () => {
  const section = { name: 'section is new' };
  const component = shallow(<SectionEdit componentKey={NEW_ELEMENT_ID} section={section} />);
  expect(toJson(component)).toMatchSnapshot();
});
test('SectionEdit shows an existing section correctly', () => {
  const componentKey = 22;
  const section = {
    name: 'section is old',
    id: componentKey,
    partTypes: ['pt1'],
  };
  const component = shallow(<SectionEdit componentKey={componentKey} section={section} />);
  expect(toJson(component)).toMatchSnapshot();
});
test('SectionEdit shows an existing section with partTypes expanded correctly', () => {
  const componentKey = 22;
  const section = {
    name: 'section is old',
    id: componentKey,
    partTypes: ['pt1', 'pt2'],
    _detail: true,
  };
  const component = shallow(<SectionEdit componentKey={componentKey} section={section} />);
  expect(toJson(component)).toMatchSnapshot();
});
