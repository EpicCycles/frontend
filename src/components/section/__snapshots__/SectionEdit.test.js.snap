// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SectionEdit shows a NEW section correctly 1`] = `
<Fragment>
  <div
    className="grid-item grid-item--fixed-left"
  >
    <AddLink
      addFunction={[Function]}
      addObjectName="Section"
    />
  </div>
  <div
    className="grid-item"
  >
    <FormTextInput
      className=""
      dataType="text"
      error=""
      fieldName="name_new"
      inputClass=""
      label=""
      list=""
      onChange={[Function]}
      onClick={[Function]}
      placeholder="add new"
      size={30}
      title=""
      value="section is new"
    />
  </div>
</Fragment>
`;

exports[`SectionEdit shows an existing section correctly 1`] = `
<Fragment>
  <div
    className="grid-item grid-item--fixed-left"
  >
    <Icon
      as="i"
      name="toggle right"
      onClick={[Function]}
    />
  </div>
  <div
    className="grid-item"
  >
    <FormTextInput
      className=""
      dataType="text"
      error=""
      fieldName="name_22"
      inputClass=""
      label=""
      list=""
      onChange={[Function]}
      onClick={[Function]}
      placeholder="add new"
      size={30}
      title=""
      value="section is old"
    />
  </div>
</Fragment>
`;

exports[`SectionEdit shows an existing section with partTypes expanded correctly 1`] = `
<Fragment>
  <div
    className="grid-item grid-item--fixed-left"
  >
    <Icon
      as="i"
      name="toggle down"
      onClick={[Function]}
    />
  </div>
  <div
    className="grid-item"
  >
    <FormTextInput
      className=""
      dataType="text"
      error=""
      fieldName="name_22"
      inputClass=""
      label=""
      list=""
      onChange={[Function]}
      onClick={[Function]}
      placeholder="add new"
      size={30}
      title=""
      value="section is old"
    />
    <PartTypes
      handleSectionChange={[Function]}
      partTypes={
        Array [
          "pt1",
          "pt2",
        ]
      }
      sectionKey={22}
    />
  </div>
</Fragment>
`;
